[WIP] Mypy suggested changes based on use of --strict-optional

neiljp opened the pull request.

smarx commented:
*Automated message from Dropbox CLA bot*

@neiljp, it looks like you've already signed the Dropbox CLA. Thanks!

gnprice commented on zerver/context_processors.py:
```diff
@@ -56,19 +56,19 @@ def zulip_default_context(request):
+    realm_name = None
+    realm_icon = None
+    realm_description = None
+    realm_invite_required = False
```
Hmm. This structure feels less nice to read than the existing code. I guess the alternative is to add type annotations on the things that are optional -- like
```
  realm_name = realm.name  # type: Optional[str]
  realm_icon = get_realm_icon_url(realm)  # type: Optional[str]
```
which isn't terrible, though it gets a little annoying on `realm_description`. Not sure I have a preference between these, though.

gnprice replied:
Huh, or we could swap the order -- put the `None` case first.

I think I prefer that over pulling it out of the conditional altogether. Feels a little clearer how there are two cases here. I'd rather keep the `None` case first, and it seems kind of awkward that mypy accepts one but not the other, but we can take that up separately.

neiljp replied:
I was following this approach after raising it in #5730 and receiving this comment from @timabbott : https://github.com/zulip/zulip/pull/5730#issuecomment-320053236

timabbott replied:
Yeah, I think in this case, with a batch of several, I definitely prefer the two parallel cases.  Swapping the order seems fine.

neiljp replied:
Done now or in upcoming revised commit.

gnprice commented on tools/test-backend:
```diff
@@ -123,7 +123,7 @@ def block_internet():
 
     httplib2.Http.request = internet_guard
-    requests.request = internet_guard
+    requests.request = lambda method, url, **kwargs: internet_guard(method, url, **kwargs)
```
Huh, interesting that this gives an error. Seems like the signature `internet_guard` already has is as general as possible, except for the return type (which doesn't change here) -- it ought to be usable where any other function that returns `None` is usable.

If `--strict-optional` causes an error here where there wasn't one already, then that's a regression in `--strict-optional` and we should pursue it upstream. In the meantime we can have a `type: ignore` here, with a comment mentioning the upstream bug.

neiljp replied:
This does seem to be the case. I've just tried to narrow down the issue to:
```
import requests

from typing import Any

def internet_guard(*args, **kwargs):
    # type: (*Any, **Any) -> None
    raise Exception()

requests.request = internet_guard
```
This does cause a problem, since `mypy` or `mypy -2` is fine, but with --strict-optional, I get:
```
11: error: Incompatible types in assignment (expression has type Callable[[VarArg(Any), KwArg(Any)], None], variable has type Callable[[str, str, KwArg(Any)], Response])
```
Does that look like a reasonable test-case?

(In addition, if I include httplib2 with the above, I need --ignore-missing-imports (as per run-mypy), since httplib2 doesn't have a typeshed (cf requests). So the former might also have complained!)

gnprice replied:
Yeah, that looks like a great test-case.

neiljp replied:
Right, I'll file a bug in mypy.

gnprice commented on tools/run-mypy:
```diff
@@ -42,8 +42,8 @@ parser.add_argument('--no-disallow-untyped-defs', dest='disallow_untyped_defs',
-parser.add_argument('--strict-optional', dest='strict_optional', action='store_true', default=False,
-                    help="""Use the --strict-optional flag with mypy""")
+parser.add_argument('--no-strict-optional', dest='strict_optional', action='store_false', default=True,
+                    help="""Don't use the --strict-optional flag with mypy""")
```
This part is the major WIP part that's not to be merged yet, I take it. :-)

neiljp commented:
I've rebased and updated the internet_guard patch to the old code with an ignore, pending discussion and/or when I file a bug in mypy.

neiljp commented:
Some outstanding issues on which I need discussion/feedback:
* `tools/setup/emoji/build_emoji`: as per a comment in zulip, should `code_point_to_file_name_map()` be using `xml.findall` rather than `xml.find`?
* the result of `get_realm` comes up in lots of places, which currently is annotated with returning Realm; as I understand it, the django query could generally be `None`, so `Optional[Realm]` makes sense, but that change involves potentially adding lots of checks/type-updates elsewhere!

timabbott commented on tools/lib/html_branches.py:
```diff
@@ -24,8 +24,8 @@ class HtmlTreeBranch(object):
-    def __init__(self, tags, fn):
-        # type: (List[TagInfo], str) -> None
+    def __init__(self, tags, fn):  # FIXME: fn is not really used and could be removed?
+        # type: (List[TagInfo], Optional[str]) -> None
```
@showell @adnrs96 can comment on whether this parameter is necessary.

adnrs96 replied:
this parameter is used in `/tools/lib/html_grep.py` basically while grepping it is used to print filenames a particular html branch belongs to.
I guess we should leave it there.

neiljp replied:
Agreed; thanks for the feedback. Removed this FIXME.

timabbott commented on zerver/models.py:
```diff
@@ -1682,12 +1681,11 @@ def to_dict(self):
         elif status == 'idle':
             status_val = UserPresence.IDLE
-        else:
-            status_val = None
```
I don't think this 3-clause if statement going to a 2-clause one makes the code more readable.  Would either annotate as optional and/or report a mypy bug that it should be able to handle this sort of situation and tag the block as `Optional[foo]`..

neiljp replied:
I believe I've seen discussion of having this situation resolve to a Union[x, y, ...] based on the result for each branch of the conditional, but my understanding is this is not fixed in mypy, though possibly related to python/mypy#2611. In the revised commit, I've marked the first branch as `Optional[int]` with link to that issue.

gnprice replied:
@neiljp , are you referring to a version of this newer than 09cf45499? It sounds like you are but that's the latest version I see on GitHub, so just wanted to make sure if there's another push you'd meant to do. :-)

neiljp replied:
This should be part of an updated commit named 'mypy: Reorder and annotate variables around if statements.'

New commits incoming soon.

timabbott commented on tools/test-backend:
```diff
@@ -123,7 +123,7 @@ def block_internet():
 
     httplib2.Http.request = internet_guard
-    requests.request = internet_guard
+    requests.request = internet_guard  # type: ignore # FIXME file bug in mypy repo!
```
I think this actually should be a `# type: ignore # monkey-patching`, since we are blatantly replacing  afunction with a different type?.

neiljp replied:
Comment adjusted for now. Shouldn't we be able to get a matching function signature, though?

gnprice replied:
@timabbott See https://github.com/zulip/zulip/pull/6017#discussion_r131818592 for previous discussion -- because the signature of `internet_guard` is already super general, I think this is actually a bug in mypy we should file.

(Not a blocker for merging, though -- now that we've had some discussion, I can just tweakt the comment as appropriate on merge.)

timabbott commented on zerver/views/home.py:
```diff
@@ -75,7 +75,7 @@ def approximate_unread_count(user_profile):
-    # type: (UserMessage) -> Optional[float]
+    # type: (Optional[UserMessage]) -> Optional[float]
     # user_message is a UserMessage object.
     if not user_message:
```
I'd write this as `if user_message is None` to follow the style of the rest.

neiljp replied:
Definitely better, yes; also removed the comment beneath, which is superfluous since we have the annotation now! A case of gaining confidence of how much to put into commits - and how much to allow "unrelated" changes, eg. extra comments or code changes!

timabbott commented on zerver/migrations/0077_add_file_name_field_to_realm_emoji.py:
```diff
@@ -125,7 +125,7 @@ def upload_to_s3(self, path, file_data, headers):
 
     def upload_files(self, response, resized_image, dst_path_id):
-        # type: (Response, binary_type, Text) -> None
+        # type: (Response, Optional[binary_type], Text) -> None
```
I'm not convinced this is the right fix; should `resize_emoji` really return `None`?

gnprice commented on zerver/models.py:
```diff
@@ -282,7 +282,7 @@ def get_realm(string_id):
 
 def completely_open(realm):
-    # type: (Realm) -> bool
+    # type: (Optional[Realm]) -> bool
```
Let's fix the style of the check just under this to `is not None`, while we're fixing `None`-related issues in this bit of code.

neiljp replied:
`is None`, right? Updated pending push.

gnprice replied:
Ah yes indeed, thanks :)

gnprice commented on zerver/models.py:
```diff
@@ -208,7 +208,7 @@ def __unicode__(self):
     @cache_with_key(get_realm_emoji_cache_key, timeout=3600*24*7)
     def get_emoji(self):
-        # type: () -> Dict[Text, Optional[Dict[str, Iterable[Text]]]]
+        # type: () -> Dict[Text, Dict[str, Iterable[Text]]]
```
Hmm, when I go consult `get_realm_emoji_uncached` to check this, taking out the `Optional` is right -- but it looks like there's a different issue, in that that return type is `Dict[Text, Dict[str, Any]]`. So effectively this annotation is making a promise that those inner values will be `Iterable[Text]`, but looking at the implementation it doesn't seem like that's true.

gnprice commented on zproject/backends.py:
```diff
@@ -414,7 +414,7 @@ class ZulipLDAPAuthBackend(ZulipLDAPAuthBackendBase):
         # type: (Text, str, Optional[Text], Optional[Dict[str, Any]]) -> Optional[UserProfile]
         try:
-            if settings.REALMS_HAVE_SUBDOMAINS:
+            if settings.REALMS_HAVE_SUBDOMAINS and realm_subdomain is not None:
```
Ah, I merged this hunk to master but was too hasty in doing so -- it actually breaks a test, like so:
https://travis-ci.org/zulip/zulip/jobs/262921012
So I've reverted that. This looked right semantically, but clearly there's more investigation needed.

Your original commit may disappear when you rebase, in which case you might cherry-pick to get it back.

neiljp replied:
This certainly makes me appreciate the tests!

After another look, perhaps the issue is simply that the function depends on `settings` to know whether to just use `get_realm`, which is an independent factor from the `realm_subdomain` parameter. On that basis, perhaps the right approach might be to move the code into the first branch?
```
if settings.REALMS_HAVE_SUBDOMAINS:
    if realm_subdomain is None:
        raise SomeException("REALMS_HAVE_SUBDOMAINS set, but no subdomain supplied")  # or return None?
    self._realm = get_realm(realm_subdomain)
else:
    self._realm = get_realm_by_email_domain(username)
```

timabbott replied:
This is a code path that's likely to be significantly restructured this week, so might not be worth working on right now.

neiljp replied:
Right, will hold off on this for now.

neiljp commented:
Commits updated, with [WAIT] annotations for those which are not yet ready to be merged/re-reviewed.

neiljp commented:
Just added a commit which potentially changes behaviour, but seems like what might have originally been intended; xml.find doesn't provide an iterable but rather a single item.

neiljp commented:
Waiting on discussion/commit before progressing with other commits here.

gnprice commented on zerver/views/home.py:
```diff
@@ -75,9 +75,8 @@ def approximate_unread_count(user_profile):
 
 def sent_time_in_epoch_seconds(user_message):
-    # type: (UserMessage) -> Optional[float]
-    # user_message is a UserMessage object.
```
:-D I appreciate when ad-hoc type annotations like this can just go away.

(In this case it already could have with the existing annotation; presumably we just missed it then.)

neiljp replied:
Yes, I really appreciate annotation 👍 
(admittedly, coming from a C++ background)

gnprice commented on tools/update-authors-json:
```diff
@@ -76,6 +76,7 @@ def run_production():
         contribs = []
         for user in json_data:
             author = user.get('author')
+            assert author is not None  # Each list entry should have an "author" dict
```
Hmm. The existing code around this looks kind of dubious, and I'm not 100% sure what error cases are actually expected.

Just below this, we have a try/except when we try to get the fields within the `author` dict... but it's not checking for `KeyError`. It's checking for `AttributeError`. That wouldn't happen if `author` just doesn't have those fields, but it *would* happen if `author` itself is `None`.

So I think the edit here that would preserve the current error behavior is to instead of an assert, write
```
if author is None:
  print("Unable ..."...)
  continue
```
Then cut the `except` block. I'm still not 100% sure that this is a situation that actually happens, and not just someone being paranoid, but we'd be avoiding a regression if it is. What do you think of that solution?

neiljp replied:
LGTM. Pushed.

gnprice commented on tools/setup/emoji/build_emoji:
```diff
@@ -144,7 +144,7 @@ def code_point_to_file_name_map(ttx):
     result = {}  # type: Dict[int, Union[Text, bytes]]
     xml = ET.parse(ttx)
-    for elem in xml.find("*cmap_format_12"):
+    for elem in xml.iterfind("*cmap_format_12"):
```
I think what's going on here is that:
* `xml.find(...)` gets us one element, and then
* we iterate through *that* element. Which gives us its child elements.

This isn't super clear, and I'd be glad to see an edit to make it clearer that that's what's happening -- even just adding something like `cmap = xml.find("*cmap_format_12")` and then `for elem in cmap:` is probably an improvement.

gnprice commented:
> zerver/tests/test_logging_handlers.py: the remaining issue with the 7th argument is I believe a typeshed issue (at least on line 83), related to mypy issue 1178, but I've not yet been able to place a type: ignore appropriately to resolve the issue.

Hmm, I see -- [the relevant line](https://github.com/python/typeshed/blob/4491e415f/stdlib/3/sys.pyi#L120) apparently says
```
# TODO should be a union of tuple, see mypy#1178
def exc_info() -> Tuple[Optional[Type[BaseException]],
                        Optional[BaseException],
                        Optional[TracebackType]]: ...
```
when it really should be `Optional[Tuple[...]]`.

For placing the `type: ignore`, the clue is in the error message -- if you look closely, it gives the line number of the function call, not of the argument or the close-paren of the function call. Because `type: ignore` is line-oriented, it goes on the line where the error message is. So this works:
```
        record = self.logger.makeRecord( # type: ignore
            'name', logging.ERROR, 'function', 16,
            'message', {}, exc_info)
```

This should be an issue in the typeshed repo -- it's not immediately obvious to me why python/mypy#1178 means it can't be fixed. Probably the immediate result of filing the issue will be to find out why that's the case :), but then at least that's on record.

neiljp commented:
A rebase and little more work here:
* trello (board_actions.py) fix
* fix in new mention.py
* ignore use of formataddr and parseaddr (for email) as the typeshed doesn't allow Optional (typeshed PR in-progress)

Unsure (no commits):
* Addressee 75/80/99: are topics/streams Optional, or not? :) (@showell ?)

neiljp commented:
@gnprice 
If you'd like to dig into these again, I'm at a stopping point for a bit.
(I may look into the issue you raised previously separately, and I'm awaiting feedback on the typeshed changes to email.utils)

timabbott commented on zerver/lib/management.py:
```diff
@@ -57,10 +57,10 @@ def add_user_list_args(self, parser, required=False, help=None, all_users_arg=Tr
         val = options["realm_id"]
         if val is None:
-            return None
+            raise CommandError("No realm specified. Aborting.")
```
We don't want to make this change; the intended UX is that you don't need to pass a realm.

gnprice commented on zerver/views/messages.py:
```diff
@@ -971,7 +971,7 @@ def send_message_backend(request, user_profile,
         (valid_input, mirror_sender) = \
             create_mirrored_message_users(request, user_profile, message_to)
-        if not valid_input:
+        if not valid_input or mirror_sender is None:  # FIXME: Amend to use exceptions?
```
Yeah, having `create_mirrored_message_users` raise exceptions would be a reasonable way to clean this up. The main downside is that we'd have to either repeat the "Invalid mirrored message" string at each of the error paths there, or have them raise one exception and then this code catch that to raise another, the `json_error`.

An alternate way would be to have `create_mirrored_message_users` return just one value, an `Optional[UserProfile]`. As it is, we have `mirror_sender` non-None just if `valid_input` is True anyway, so we could cut the redundancy.

timabbott replied:
Yeah, I like that approach.

gnprice commented on zerver/lib/mention.py:
```diff
@@ -28,6 +28,6 @@ def extract_name(s):
     matches = re.findall(find_mentions, content)
-    names = {extract_name(match) for match in matches}
-    names = {name for name in names if name}
+    names_with_none = {extract_name(match) for match in matches}
```
This code bugged me in that it makes a whole `set` before throwing it away to make nearly the same `set`. Sadly Python's comprehensions lack a way to write these cleanly as one comprehension, but I ended up adding a tiny tweak here to make `names_with_none` a genexp.

neiljp commented:
@gnprice I removed the commit that @timabbott referenced and have substituted assertions (where the realm is required in the parser, so should be enforced that way?) or adjusted conditionals where appropriate to depend upon the results of `get_realm` rather than the input.

That should return the situation to be about 50 remaining 🎉

gnprice commented:
Thanks again @neiljp! Just pushed (ab4cbf81f and ancestors) a slightly tweaked version of those two commits: there was one command (fix_unreads) where the option isn't actually required (`add_realm_args` is called without passing `required=True`), so the assert isn't right but I adjusted a type annotation instead.

Also in soft_activate_deactivate_users, I tweaked the conditionals to use "if not None", and also took the opportunity to simplify some of the related logic.

neiljp commented:
The formataddr/parseaddr issue has been merged into typeshed via python/typeshed#1626.

Do we wait for mypy to sync typeshed before upgrading to another mypy git release, and then drop that commit? (assuming the need for it does disappear!)

gnprice commented:
​Great!

Yeah, waiting for mypy to sync typeshed is generally easiest -- they
typically do that multiple times a week (as of when I last paid close
attention).

timabbott commented:
@neiljp are any pieces of this things we should be trying to merge now before we start switching to the Python 3 syntax?

timabbott commented:
@neiljp good news!  The work on refactoring all of our auth backends/realm logic is at this point done, so if you'd like to do a pass through cleaning those up, that should now be doable.

neiljp commented:
I've rebased and cleaned up some of these and added some more, so they should be able to be merged at least now.

@gnprice I'm guessing you'll take the lead on this now?

showell commented:
@neiljp What is the current status of this PR?  I think you're likely to keep getting conflicts from GCI students fixing stuff, so it might be good to stay on top of rebasing this.

gnprice commented:
I just noticed this PR again, actually because of one of those conflicts -- because it caused zulipbot to make a comment, which generated an email notification for me :stuck_out_tongue: 

I've just merged all the commits except the WAIT commits and one with FIXME. Thanks again @neiljp !

For the sake of tracking our progress here in this PR thread: with just the WAIT commits of this branch (which set up `--strict-optional`), `tools/run-mypy` printed 181 lines of errors. With the whole branch, it printed 171.

(I tested that on not quite the latest master, so the numbers may vary slightly.)

neiljp commented:
Rebased & a few more commits (5 fewer errors, running mypy on a droplet).

neiljp commented:
If the above commits are acceptable, this should take the outstanding mypy error total down to 113.

neiljp commented:
The tests are currently failing on coverage checks, based on two code changes I added:
* extra check for `None`at start of `integration_doc`; can `integration_name` just be `str`, not `Optional`?
* extracting the `mdiff.diff_strings` above the `raise AssertionEror` in `BugdownTest.assertEqual`; I'm unsure how that is tested...and it's a test, after all?

gnprice commented on zerver/views/integrations.py:
```diff
@@ -131,7 +131,10 @@ def get_context_data(self, **kwargs: Any) -> Dict[str, Any]:
 @has_request_variables
-def integration_doc(request: HttpRequest, integration_name: str=REQ(default=None)) -> HttpResponse:
+def integration_doc(request: HttpRequest,
+                    integration_name: Optional[str]=REQ(default=None, type=str)) -> HttpResponse:
```
Yeah, I'm pretty sure that from the pattern in `zproject/urls.py` that points here, this has to be non-None.

neiljp replied:
OK, I've adjusted this. For future reference, is it obvious what can lead to `None`?

gnprice replied:
Sure. This URL route has a group for `integration_name` right in the regex for the route:
```
    url(r'^integrations/doc-html/(?P<integration_name>[^/]*)$',
        zerver.views.integrations.integration_doc,
        name="zerver.views.integrations.integration_doc"),
```
And that group has to be there for the pattern to match -- it's not followed by a `?`, or on one side of a `|`, or affected by any other regex operator that could let the overall pattern match without having that bit match something.

So, then, whatever matches that group gets supplied as the argument `integration_name`.

There are some other routes where the group is optional:
```
    url(r'^accounts/send_confirm/(?P<email>[\S]+)?',
        TemplateView.as_view(template_name='zerver/accounts_send_confirm.html'),
        name='send_confirm'),
```
and we also often have routes with arguments that aren't in the URL path at all, but in say the POST body. Then those can be missing and the URL router will still call the underlying function to handle it, passing None.

gnprice commented on zerver/tests/test_bugdown.py:
```diff
@@ -217,8 +217,9 @@ class BugdownTest(ZulipTestCase):
-                                     mdiff.diff_strings(first, second) + msg)
+                diff = mdiff.diff_strings(first, second)
+                raise AssertionError("Actual and expected outputs do not match; showing diff.\n%s%s" %
+                                     (diff if diff is not None else "[Could not perform diff]", msg))
```
Hmm, it looks like we only use `mdiff` in tests.

Rather than have it return None in its error case, let's have it just raise an exception. That seems like cleaner behavior anyway -- after all, that's what will happen if the subprocess fails in some unanticipated way -- and there's little risk to worry about in changing it, because it is only in tests.

neiljp commented:
Rebased; commits ready for review once more.

neiljp commented:
@gnprice Rebased FYI.

timabbott commented:
@neiljp can you give this another rebase?  I think it has some merge conflicts with some 2to3 syntax changes.

gnprice commented:
Thanks again @neiljp ! I've just merged nearly all of the non-WAIT commits, as usual; 12 commits in all.
* The two "Use REQ type parameter" commits had several merge conflicts, as Tim said above; probably not hard for you to resolve.
* I wasn't a fan of adding a bunch of casts, because I'm pretty sure semantically they had the same meaning as giving names (and types) to those keyword arguments, and the latter syntax seems much clearer. And happily, when I tried that, it all worked! So I merged that instead, as 7bfd22a0b.
* get_realm_emoji_uncached: `Dict[Text, Optional[foo]]` seems like an unusual type -- do we really add Nones as values to this dict? We don't right here, but the dict gets returned and of course dicts are mutable, so maybe we do elsewhere. I'd want to be confident this type really describes what's happening before writing it down.

gnprice commented:
Continuing our tracking: as of when this PR branch was last rebased, Jan 12, it reduced the `--strict-optional` error line count from 199 to 135.

Rebased to a current master, the error line count started at 246 and these changes reduced it to 219. So, there's some backsliding happening in master.

Looking at the 89 lines that appeared (between the tip of the PR branch, and the new master I merged plus the WAIT setup commits):
* ~35 are `error: "REQ" does not return a value` -- about half in webhooks, half elsewhere. Hmm, actually I guess these are probably from merging your preparatory REQ changes but not the commits that used that. So after you rebase those changes and I merge them, that will probably fix these.
* ~30 more are in tests, passing an `Optional[UserProfile]` to something that expects `UserProfile`
* misc others

It feels like we're getting close, but of course if people are regularly creating new errors because they know about them, it's going to get to be slower going the closer we get.

What would you think about sending a PR to start to lock this down for the files that are clean? The ingredients to do this would be
* Mypy has [a config file](http://mypy.readthedocs.io/en/latest/config_file.html).
* One of the options is `strict_optional`, and it has exactly the same effect as the command-line flag. The docs say it's global, but in fact it works as a per-file option too.
* As an example, [here's the change](https://github.com/python/mypy/pull/4070/files#diff-bda2dfcf72739d7d683f00aa4b4d70f9) where mypy itself removed the last exceptions to checking its own codebase with `--strict-optional`.

For us, I think this PR would look something like
* start a `mypy.ini` file
* while we're at it, maybe move some of the miscellaneous stuff in `run-mypy` into that file (this part is optional)
* start writing some per-file patterns in there that turn on strict-optional for parts of the codebase, keeping everything passing
* maybe even flip the default already: set `strict_optional = True` in the global section, and turn it off for specific file patterns.

How's that sound?

timabbott commented:
I had been having the same thought with `mypy.ini`.  Especially since we have a few files like `queue.py` where strict-optional may be very difficult.

neiljp commented:
@gnprice First stage: a rebase & drop of the casting commit.
Next: look at the emoji file & then consider the other PR.

neiljp commented:
Emoji issue resolved. Down to 182 errors locally.
@gnprice FYI.

neiljp commented:
A few more updates, taking the count to 148.

timabbott commented on zerver/webhooks/appfollow/view.py:
```diff
@@ -15,15 +15,17 @@
 def api_appfollow_webhook(request: HttpRequest, user_profile: UserProfile,
                           stream: Text=REQ(default="appfollow"),
-                          topic: Optional[Text]=REQ(default=None),
+                          topic: Optional[Text]=REQ(default=None, type=str),
```
So, with `REQ` with no special arguments, the type should always be `str`.  Would it be better to adjust `REQ`, rather than changing every call point to add `type=str`?

timabbott commented on zerver/lib/test_classes.py:
```diff
@@ -327,12 +327,14 @@ def submit_reg_form_for_user(
         for message in reversed(outbox):
             if email_address in message.to:
-                return re.search(url_pattern, message.body).groups()[0]
+                return re.search(url_pattern_to_use, message.body).groups()[0]
```
This hunk feels a bit messy; is there a better way to do this?

neiljp replied:
Do you just mean this commit? Messy how?

The issue in this commit is that `url_pattern`  is `Optional[Text]`, whereas `re.search` accepts an `AnyStr`. I didn't want to rename the function parameter, so introduced another name internally; a different internal one would be fine if it read better. I also reordered the conditional - is that the particular part that seems messy?

timabbott commented on zerver/views/user_settings.py:
```diff
@@ -126,13 +126,13 @@ def json_change_settings(request: HttpRequest, user_profile: UserProfile,
+        twenty_four_hour_time: Optional[bool]=REQ(validator=check_bool, default=None, type=bool),
+        high_contrast_mode: Optional[bool]=REQ(validator=check_bool, default=None, type=bool),
+        night_mode: Optional[bool]=REQ(validator=check_bool, default=None, type=bool),
+        default_language: Optional[bool]=REQ(validator=check_string, default=None, type=bool),
```
This one seems wrong; should be `type=str`.

neiljp replied:
The `default_language` change was to match the parameter annotation, which was a previous edit by yourself (03b4a024) where the number of annotations in python2 style didn't match the number of parameters. So, if the parameter should be `Optional[str]` then `type=str` makes sense too.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -3497,7 +3497,7 @@ def user_info(um: UserMessage) -> Dict[str, Any]:
 def do_update_message(user_profile: UserProfile, message: Message, topic_name: Optional[Text],
                       propagate_mode: str, content: Optional[Text],
-                      rendered_content: Optional[Text], prior_mention_user_ids: Set[int],
+                      rendered_content: Text, prior_mention_user_ids: Set[int],
```
`rendered_content` should be optional; the structure of this function is that you could have changed either the topic or the content (or both), so we can't assume either is not `None`.  

The correct inference is "content is not None" means "rendered_content is not None".

neiljp replied:
That's fine; this change resolved a `None` value of `rendered_content` being used, so that'll be there for another day instead :)

timabbott commented:
I merged about a half dozen commits and posted comments on a few things I'm worried about.

neiljp commented:
I'll wait a bit before pushing a rebase, to allow for discussion; in the meantime, is there a reason we're waiting on at least 'Update Optional returns in zerver/models.py'? I'm happy to change the commit title for it, since it seems a little vague, but the code seems ok?

timabbott commented:
For merging something like `Update Optional returns in zerver/models.py`, it'd be good to have a commit title that fascilitates review.

@neiljp I think this PR has gotten kinda unwieldy (really long comment log above).  Would you be up for rebasing, and submit one or more new PRs that just has pieces that aren't merged yet, with the comments addressed? I think that sort of new PR approach would be much more reviewable at this point.

neiljp commented:
@timabbott We merged in the `mypy.ini` changes, which makes piecewise addition of the remaining problems a lot easier, as per your and Greg's suggestions :)

I had left this open for a bit, since when I had some time previously I was looking at some previous components which hadn't been merged, which lead to #8904.
